### PR TITLE
[bitnami/mariadb-galera] Fix bootstraping issues

### LINKED ISSUE #2810: [bitnami/mariadb-galera] New bootstrap options break upgrade/rollout
**Which chart**:
mariadb-galera, 3.1.3 (since 3.0.0)

**Describe the bug**
The chart version 3.x introduced the force bootstrap feature: https://github.com/bitnami/charts/pull/2727
This is really a needed feature for recovery from a failure state, but it introduced a major regression: A rolling restart of the mariadb-galera pods does not work anymore, which is performed e.g. by a chart upgrade. By default, it will always bootstrap from the pod with index 0. If the cluster is up and running (with pod 1 and 2 that were already upgraded by the rolling restart), pod 0 will want to bootstrap anyway. Of course, safe_to_bootstrap is not set for pod 0 (and it should not be). Therefore it will not start with the error message `It is not safe to bootstrap form this node ('safe_to_bootstrap=0' is set in 'grastate.dat')` and be stuck in a crash loop.

**To Reproduce**
1. Deploy a new mariadb-galera cluster with the default of 3 nodes and persistence
2. Simulate a rollout of a newer chart version with "kubectl rollout restart statefulset ..."
3. Observe that pod 2 is restarted successfully, it joins the cluster
4. Observe that pod 1 is restarted successfully, it joins the cluster
5. Observe that pod 0 tries to restart, but fails with `It is not safe to bootstrap form this node ('safe_to_bootstrap=0' is set in 'grastate.dat')` 

**Expected behavior**
* Bootstrap should not be a default option. It introduces the risk of pods that refuse to start (even though they could join the cluster) or even having two "clusters" (split brain).
* If a cluster already is up and running, a new node should never bootstrap. It's better to fail (it does that currently only when persistence is active).
* Nonetheless, it would of course be nice that on initial deployment, the initial bootstrap works. The chart should still be easy to use.

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.0.2", GitCommit:"19e47ee3283ae98139d98460de796c1be1e3975f", GitTreeState:"clean", GoVersion:"go1.13.5"}
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:52:00Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.6", GitCommit:"d32e40e20d167e103faf894261614c5b45c44198", GitTreeState:"clean", BuildDate:"2020-05-20T13:08:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Additional context**

**Workaround**
If the cluster is functional and no bootstrap is necessary, you can set `galera.bootstrap.bootstrapFromNode` to some large value (larger than the number of nodes). Then no bootstrap wil be tried.

**Broken current implementation**
The current implementation uses an EmptyDir volume with a `.bootstrap/done` file to avoid the bootstrap on a pod restart. Unfortunately, this is not enough as a rollout creates new pods. This may be a bit surprising because a StatefulSet is used. But just thinking about the case that the pod could be scheduled on another node (with the same data volume), this makes sense.

**Warning for persistence=false**
When persistence is deactivated (e.g. for testing), this issue may not be apparent. The reason is that the first node will be able to bootstrap with an empty database - always! So there will be two clusters, one with `wsrep_cluster_size=1` and one with `wsrep_cluster_size=2`. That's even worse.

### PR DESCRIPTION
**Pull request by rafariossaa**
**Description of the change**
This PR fixes bootstrap issues when for example rolling out an upgrade like in #2810.

**Benefits**

Remove the need of workaround like using a high `bootstrapFromNode` value.

**Possible drawbacks**

**Applicable issues**

  - fixes #2810 

**Additional information**

**Checklist** 
- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [-] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [X] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

:warning: Keep in mind that if you want to make changes to the kubeapps chart, please implement them in the [kubeapps repository](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps). This is only a synchronized mirror.

### CODE REVIEW

File: bitnami/mariadb-galera/values-production.yaml [hunk 1]
```diff
@@ -182,13 +182,14 @@ galera:
...
   bootstrap:
     ## Node to bootstrap from, you will need to change this parameter incase you want to bootstrap from other node
     ##
-    bootstrapFromNode: 0
-    ## Force bootstrap
+    bootstrapFromNode:
+    ## Force safe_to_bootstrap in grastate.date file.
     ## This will set safe_to_bootstrap=1 in the node indicated by bootstrapFromNode.
-    forceSafeToBootstrap: false
+    # forceSafeToBootstrap: false
```
Comment (andresbono) [inline bitnami/mariadb-galera/values-production.yaml:192]: Why uncommenting this? The same in the `values.yaml`

Comment (rafariossaa) [reply bitnami/mariadb-galera/values-production.yaml]: It is true, it would be better to left them uncommeted.

**Comment by rafariossaa** [comment]
Just to share the testing done.


The image used `myrepo/galera` is the one used by the chart but adding extra tracing info to `get_galera_cluster_bootstrap_value`.

**Installation:**

```
helm install mygal  --set image.debug=true   --set image.repository=myrepo/galera  --set image.tag=latest  --set image.pullPolicy=Always  ~/projects/bitnami-charts/bitnami/mariadb-galera/.
```

**Test cases:**


1- **Just intall**:
    - Cluster status: OK, size 3
    - grastate: OK, sequno -1 as there is no stopped node, safe_to_boostrap = 0 in all nodes. The uuid matches with previous test.
    - nodes running: OK, 3 nodes running.
    - Data insertion and replication: OK, data inserted and replicated.


2- **Install, kill pod, node automatically restarted**
Pod killed with: `kubectl delete pod  mygal-mariadb-galera-0`
Tested killing the 3 nodes.

  - Cluster status: OK, size 3
  - grastate: OK, sequno -1 as there is no stopped node, safe_to_boostrap = 0 in all nodes. The uuid matches with previous test.
  - nodes running: OK, 3 nodes running.
  - Data insertion and replication: OK, data inserted and replicated.


3- **Start with previous data: This is also the case for restoring from backup.**
helm install..., helm delete, helm install forcing: This is also the case for restoring from backup.

```
   helm install mygal  --set image.debug=true --set rootUser.password=testpwd --set db.password=testpwd --set galera.mariabackup.password=testpwd --set image.repository=myrepo/galera  --set image.tag=latest  --set image.pullPolicy=Always  ~/projects/bitnami-charts/bitnami/mariadb-galera/.

```

```
   helm install mygal --set image.debug=true --set rootUser.password=testpwd --set db.password=testpwd --set galera.mariabackup.password=testpwd --set image.repository=myrepo/galera  --set image.tag=latest  --set image.pullPolicy=Always  --set galera.bootstrap.bootstrapFromNode=0 --set galera.bootstrap.forceSafeToBootstrap=true ~/projects/bitnami-charts/bitnami/mariadb-galera/.

```
  - Cluster status: OK, size 3
  - grastate: OK, sequno -1 as there is no stopped node, safe_to_boostrap = 0 in all nodes. The uuid matches with previous test.
  - nodes running: OK, 3 nodes running.
  - Data insertion and replication: OK, data inserted and replicated.

4- **Simulate upgrade:**
helm install...

```
   helm install mygal  --set image.debug=true --set rootUser.password=testpwd --set db.password=testpwd --set galera.mariabackup.password=testpwd --set image.repository=myrepo/galera  --set image.tag=latest  --set image.pullPolicy=Always  ~/projects/bitnami-charts/bitnami/mariadb-galera/.

```
upgrade:

```
   kubectl rollout restart sts/mygal-mariadb-galera
```

  - Cluster status: OK, size 3
  - grastate: OK, sequno -1 as there is no stopped node, safe_to_boostrap = 0 in all nodes. The uuid matches with previous test.
  - nodes running: OK, 3 nodes running.
  - Data insertion and replication: OK, data inserted and replicated.

5- **Upgrade:**
helm install...

```
   helm install mygal  --set image.debug=true --set rootUser.password=testpwd --set db.password=testpwd --set galera.mariabackup.password=testpwd --set image.repository=myrepo/galera  --set image.tag=latest  --set image.pullPolicy=Always  ~/projects/bitnami-charts/bitnami/mariadb-galera/.

```
helm upgrade (latest and latest_up only differ in one info message):

```
   helm upgrade mygal  --set image.debug=true --set rootUser.password=testpwd --set db.password=testpwd --set galera.mariabackup.password=testpwd --set image.repository=myrepo/galera  --set image.tag=latest_up  --set image.pullPolicy=Always  ~/projects/bitnami-charts/bitnami/mariadb-galera/.

```
  - Cluster status: OK, size 3
  - grastate: OK, sequno -1 as there is no stopped node, safe_to_boostrap = 0 in all nodes. The uuid matches with previous test.
  - nodes running: OK, 3 nodes running.
  - Data insertion and replication: OK, data inserted and replicated.

**Checks:**

- Cluster status:

```
     kubectl run mygal-mariadb-galera-client --rm --tty -i --restart='Never' --namespace rafael --image myrepo/galera:latest --command \
        -- mysql -h mygal-mariadb-galera -P 3306 -uroot -p$(kubectl get secret --namespace rafael mygal-mariadb-galera -o jsonpath="{.data.mariadb-root-password}" | base64 --decode) my_database

     SHOW GLOBAL STATUS LIKE 'wsrep_cluster_state_uuid';
     SHOW GLOBAL STATUS LIKE 'wsrep_cluster_size';
     SHOW GLOBAL STATUS LIKE 'wsrep_cluster_status';

```
 * grastate: check cluster uuid and seqno and safe_to_boostrap values with . All nodes same uuid, proper seqno and safe_to_bootstrap.

```
     kubectl exec -it mygal-mariadb-galera-0 -- cat /bitnami/mariadb/data/grastate.dat
     kubectl exec -it mygal-mariadb-galera-1 -- cat /bitnami/mariadb/data/grastate.dat
     kubectl exec -it mygal-mariadb-galera-2 -- cat /bitnami/mariadb/data/grastate.dat

```
- Nodes running:

```
     kubectl get pod

```
 - Data insertion and replication:

```
     kubectl run mygal-mariadb-galera-client --rm --tty -i --restart='Never' --namespace rafael --image myrepo/galera:latest --command \
        -- mysql -h mygal-mariadb-galera -P 3306 -uroot -p$(kubectl get secret --namespace rafael mygal-mariadb-galera -o jsonpath="{.data.mariadb-root-password}" | base64 --decode) my_database

     create table t1 (c1 varchar(100));
     insert into t1 values ("123");
     insert into t1 values ("123");
     insert into t1 values ("123");
     select count(*) from t1;


```

```
     kubectl exec -it mygal-mariadb-galera-0  -- mysql -h localhost -P 3306 -uroot -p$(kubectl get secret --namespace rafael mygal-mariadb-galera -o jsonpath="{.data.mariadb-root-password}" | base64 --decode) my_database -e "select count(*) from t1"

```

```
     kubectl exec -it mygal-mariadb-galera-1  -- mysql -h localhost -P 3306 -uroot -p$(kubectl get secret --namespace rafael mygal-mariadb-galera -o jsonpath="{.data.mariadb-root-password}" | base64 --decode) my_database -e "select count(*) from t1"

```

```
     kubectl exec -it mygal-mariadb-galera-2  -- mysql -h localhost -P 3306 -uroot -p$(kubectl get secret --namespace rafael mygal-mariadb-galera -o jsonpath="{.data.mariadb-root-password}" | base64 --decode) my_database -e "select count(*) from t1"

```

**Comment by lcfang** [comment]
Hi, @rafariossaa , this pr is so great to resolve some accident. But I have a doubt about an extreme scenario: all the there pod died because the k8s node outage,  after the power restoration, no one pod can be Running but always CrashLoopBackOff  with the error:`ERROR ==> It is not safe to bootstrap form this node ('safe_to_bootstrap=0' is set in 'grastate.dat'). If you want to force bootstrap, set the environment variable MARIADB_GALERA_FORCE_SAFETOBOOTSTRAP=yes
` . I search that , all the seqno in grastate.dat is -1, and only on pod has gvwstate.dat with 2 members, in this scenario, I think I should choose a pod as safe_to_bootstrap, but I am not sure which one I should choose.
All the grastate.dat and gvwstate.dat files are:
```
 cat mysql-data-mysql-mariadb-galera-0-pvc-876725bc-84e3-45f0-81ac-b8cd74c2175f/data/grastate.dat; echo "=====mysql-mariadb-galera-0.gvwstate.dat"; cat mysql-data-mysql-mariadb-galera-0-pvc-876725bc-84e3-45f0-81ac-b8cd74c2175f/data/gvwstate.dat; echo "=====mysql-mariadb-galera-1.grastate.dat"; cat mysql-data-mysql-mariadb-galera-1-pvc-e1d44599-da66-4c8d-87b7-e8eecb1b0a65/data/grastate.dat; echo "=====mysql-mariadb-galera-1.gvwstate.dat"; cat mysql-data-mysql-mariadb-galera-1-pvc-e1d44599-da66-4c8d-87b7-e8eecb1b0a65/data/gvwstate.dat; echo "=====mysql-mariadb-galera-2.grastate.dat"; cat mysql-data-mysql-mariadb-galera-2-pvc-211d24a4-8e78-4f02-9549-10a3101f251a/data/grastate.dat; echo "=====mysql-mariadb-galera-2.gvwstate.dat"; cat mysql-data-mysql-mariadb-galera-2-pvc-211d24a4-8e78-4f02-9549-10a3101f251a/data/gvwstate.dat; echo "==================================="; echo "==================================="; 
=====mysql-mariadb-galera-0.grastate.dat
# GALERA saved state
version: 2.1
uuid:    3fc070ad-cfb8-11ea-9133-ff2fe40b4543
seqno:   -1
safe_to_bootstrap: 0
=====mysql-mariadb-galera-0.gvwstate.dat
cat: mysql-data-mysql-mariadb-galera-0-pvc-876725bc-84e3-45f0-81ac-b8cd74c2175f/data/gvwstate.dat: No such file or directory
=====mysql-mariadb-galera-1.grastate.dat
# GALERA saved state
version: 2.1
uuid:    3fc070ad-cfb8-11ea-9133-ff2fe40b4543
seqno:   -1
safe_to_bootstrap: 0
=====mysql-mariadb-galera-1.gvwstate.dat
cat: mysql-data-mysql-mariadb-galera-1-pvc-e1d44599-da66-4c8d-87b7-e8eecb1b0a65/data/gvwstate.dat: No such file or directory
=====mysql-mariadb-galera-2.grastate.dat
# GALERA saved state
version: 2.1
uuid:    3fc070ad-cfb8-11ea-9133-ff2fe40b4543
seqno:   -1
safe_to_bootstrap: 0
=====mysql-mariadb-galera-2.gvwstate.dat
my_uuid: f6486280-cfce-11ea-aa97-ce1a9cee1838
#vwbeg
view_id: 3 e6f52b49-cfd3-11ea-903d-9bc7e3065386 26
bootstrap: 0
member: e6f52b49-cfd3-11ea-903d-9bc7e3065386 0
member: f6486280-cfce-11ea-aa96-ce1a9cee1838 0
#vwend
```
I try to set safe_to_bootstrap=1 for pod0 manually, then all the pod Running.
@rafariossaa Do you have any idea about this accident? Or some better solution for this? Look forward to your reply! Thanks again.

**Comment by andresbono** [comment]
Hi @lcfang, as none of the pods had a graceful shutdown, there is no valid seqno. You usually check the nodes looking for the biggest seqno. That indicates the node that has the most recent data.

> I try to set safe_to_bootstrap=1 for pod0 manually, then all the pod Running.

So did you resolve the issue by bootstrapping pod0? Please, open a new issue if you need it, that way we can do a better follow-up.

**Comment by lcfang** [comment]
@andresbono Thank you for your attention, I have open a new issue.
Additionally, there is another scenario like above: helm delete the repo without deleting the pvc created, and then install again, I found that the cluster can't work. Eventhough I set the node0 as safe_to_bootstrap one.

**Comment by andresbono** [comment]
Thanks for creating the new issue.

About the second scenario you are describing, it is likely that passwords are being regenerated again and it causes conflicts. You should make sure the password used on the first install is the same than in the second one. You can do that by specifying the password in the values or by using a secret.

**Comment by lcfang** [comment]
@andresbono 

> Thanks for creating the new issue.

It's  my pleasure.

> About the second scenario you are describing, it is likely that passwords are being regenerated again and it causes conflicts. You should make sure the password used on the first install is the same than in the second one. You can do that by specifying the password in the values or by using a secret.

which password should I set? I try to set rootUser.password=xxx to install the helm chart, then delete it without deleting pvc, and helm install again with rootUser.password=xxx. but it also start with error `It is not safe to bootstrap form this node ('safe_to_bootstrap=0' is set in 'grastate.dat'). If you want to force bootstrap, set the environment variable MARIADB_GALERA_FORCE_SAFETOBOOTSTRAP=yes`

**Comment by andresbono** [comment]
Hi @lcfang, we tried to reproduce the issue but we couldn't do it. To restore the cluster, we basically followed this section of the readme: [All the nodes with safe_to_bootstrap: 0](https://github.com/bitnami/charts/tree/master/bitnami/mariadb-galera#all-the-nodes-with-safe_to_bootstrap-0).

In summary:

```
helm install mariadbgalera bitnami/mariadb-galera --set rootUser.password=mypassword --set db.password=mypassword --set galera.mariabackup.password=mypassword
helm delete mariadbgalera
helm install mariadbgalera bitnami/mariadb-galera --set rootUser.password=mypassword --set db.password=mypassword --set galera.mariabackup.password=mypassword --set galera.bootstrap.bootstrapFromNode=0 --set galera.bootstrap.forceSafeToBootstrap=true
```

**Pull request merged**